src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -351,19 +351,43 @@ private bool Equals(ArrayTypeSymbol other, TypeCompareKind comparison)
                 return true;
             }
 
-            if ((object)other == null || !other.HasSameShapeAs(this) ||
-                !other.ElementTypeWithAnnotations.Equals(ElementTypeWithAnnotations, comparison))
+            // We don't want to blow the stack if we have a type like T[][][][][][][][]....[][],
+            // so we do not recurse until we have a non-array. Rather, hash all the ranks together
+            // and then hash that with the "T" type.
+            var array = this;
+            do
             {
-                return false;
-            }
+                if (other is null || !other.HasSameShapeAs(array))
+                {
+                    return false;
+                }
 
-            // Make sure bounds are the same.
-            if ((comparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0 && !this.HasSameSizesAndLowerBoundsAs(other))
-            {
-                return false;
-            }
+                // Make sure bounds are the same
+                if ((comparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0 && !array.HasSameSizesAndLowerBoundsAs(other))
+                {
+                    return false;
+                }
 
-            return true;
+                var arrayElementType = array.ElementTypeWithAnnotations;
+                var otherElementType = other.ElementTypeWithAnnotations;
+                if (arrayElementType.IsSZArray())
```

Comment:
>if (arrayElementType.IsSZArray()) [](start = 16, length = 33)

I am not sure why only SZArray gets special treatment.

Reply:
I think we should handle all arrays here.

---
In reply to: [274701743](https://github.com/dotnet/roslyn/pull/34891#discussion_r274701743) [](ancestors = 274701743)

Reply:
This thread is marked as resolved, but it doesn't look like the code was changed, we are still covering only SZArrays. 

---
In reply to: [274705339](https://github.com/dotnet/roslyn/pull/34891#discussion_r274705339) [](ancestors = 274705339,274701743)

Reply:
Sigh, I added the right helper but forgot to call it here :frown: 

---
In reply to: [275032063](https://github.com/dotnet/roslyn/pull/34891#discussion_r275032063) [](ancestors = 275032063,274705339,274701743)

Reply:
Actually fixed it this time. Will be in next push. 

---
In reply to: [275034531](https://github.com/dotnet/roslyn/pull/34891#discussion_r275034531) [](ancestors = 275034531,275032063,274705339,274701743)